Feature/refactor auth

sijumoncy opened the pull request:
**Basic Refactor of Auth module **

- [x]  Authentication Change to DB Based from File Based
- [x] update Seed Db scripts for default values and tables
- [x] Moving permission map, access rules and corresponding data to DB
- [x] Permission map, access rules and other required, default values added in csv to upload through seed script 
- [x] Implemented Functions to load required data from DB
- [x] Refactored authentication module for the changes
- [x] Necessary changes in test config and test cases
- [x] Endpoints for App ( Register, Generate Key, Remove key, Delete App, Edit )
- [x] Removed Header based app identification to key through query params. Query params for key added for all endpoints
- [x] kratos instance and schema created for handling app (only working on local, need to fix other ymls) ( using V10.1 )

Test is not running because of kratos version and issue in config yml (Working fine in local )

kavitharaju commented on app/auth/auth_app.py:
```diff
@@ -0,0 +1,214 @@
+        raise GenericException(app_update_response["error"]["status"])
+    return app_updated_data
+
+DEFAULT_APPS = [
```
Is this the place we keep this data? Shouldn't it be at seed DB?
Only password fields need to be read from ENV and filled in, in code, rest can be from seed DB, right?

sijumoncy replied:
The basic data of default apps keep in one of the code base file is because we are using this data in other places ( like test conf) .  The seed Db is only using the name of the default app not other details. we can set the password field as well as the email from the ENV.

kavitharaju commented on app/auth/auth_app.py:
```diff
@@ -0,0 +1,214 @@
+            "phone": "None"
+        }
+    },
+    {
```
We dont need a separate App for API user. If no app token is provided we can assume its API user.

sijumoncy replied:
yes the code base is working in that way , no token is provided will be treated as request from API. This is currently only used in test , we can remove this app from the default app list if needed

kavitharaju replied:
If you are planning to do this in separate PR make sure you create an issue for this or a task in a main issue

sijumoncy replied:
https://github.com/Bridgeconn/vachan-api/issues/411 - added it as a task

kavitharaju commented on app/db_models.py:
```diff
@@ -432,3 +432,83 @@ class Jobs(Base): # pylint: disable=too-few-public-methods
+
+    roleId = Column('role_id', Integer, primary_key=True)
+    roleName = Column('role_name', String, unique=True, index=True)
+    createdUser = Column('created_user', String)
```
It may be good to have role description and also a field associated App, which can be null also

sijumoncy replied:
we already have a relation with app and role . right ? app with default role . do we need the relation 2 way . 
If so does that makes complex database operations
register app :
 role add to table first for a new app
app add to app table and  that role add to default role in app table

sijumoncy replied:
WE can enforce to have role connected to an app name when add a new ROLE in db via role add API

kavitharaju commented on app/routers/auth_api.py:
```diff
@@ -155,3 +163,104 @@ async def delete_user(user:schema_auth.UserIdentity,request: Request,#pylint: di
+    responses={400: {"model": schemas.ErrorResponse},422: {"model": schemas.ErrorResponse},
+    500: {"model": schemas.ErrorResponse}, 409: {"model": schemas.ErrorResponse}},
+    status_code=201,tags=["App"])
+# @get_auth_access_check_decorator
```
why is the decorator commented off?

sijumoncy replied:
Forgot to remove comment , commented on development

kavitharaju commented on app/routers/auth_api.py:
```diff
@@ -155,3 +163,104 @@ async def delete_user(user:schema_auth.UserIdentity,request: Request,#pylint: di
+    * Successful verifiacation will return a key for app'''
+    log.info('In key Login')
+    log.debug('login:%s',app_email)
+    return login_kratos(app_email,password,from_app=True)
```
If we use the same function as for users, then where do we tell that its supposed to use the other Kratos?

kavitharaju commented on app/routers/auth_api.py:
```diff
@@ -155,3 +163,104 @@ async def delete_user(user:schema_auth.UserIdentity,request: Request,#pylint: di
+responses={403: {"model": schemas.ErrorResponse},404:{"model": schemas.ErrorResponse},
+401: {"model": schemas.ErrorResponse}}
+,tags=["App"])
+def delete_key(app_key: types.SecretStr,
```
Auth decorator is missing

kavitharaju commented on app/routers/content_apis.py:
```diff
@@ -24,8 +25,9 @@
-     user_details =Depends(get_user_or_none),db_: Session = Depends(get_db)):
+    app_key: types.SecretStr = Query(None),#pylint: disable=unused-argument
+    skip: int = Query(0, ge=0),limit: int = Query(100, ge=0),
+    user_details =Depends(get_user_or_none),db_: Session = Depends(get_db)):
```
Wouldn't we need a dependency, app_details = Depends(get_app_or_none) ? Of it is it handled inside the auth decorator>

sijumoncy replied:
user details dependency is for the user who is requesting ?

kavitharaju reviewed (commented):
Reviewed only the files with core changes. But lets continue the discussions on this PR.

kavitharaju commented on app/schema/schema_auth.py:
```diff
@@ -43,15 +42,15 @@ class EditUser(BaseModel):
+#     AGUSER = 'AgUser'
+#     VACHANUSER = 'VachanUser'
+#     APIUSER = 'APIUser'
+#     BCSDEV = 'BcsDeveloper'
```
We could remove the commented portions. Coming back and cleaning up may not happen always. If we need to reference them we will be keeping the version-2 branch seperate, right? So no risk I guess.

sijumoncy replied:
ok, will remove these

kavitharaju commented on app/schema/schema_auth.py:
```diff
@@ -63,7 +62,7 @@ class FilterRoles(str, Enum):
     """kratos user role input"""
     userid:str
-    roles:List[AdminRoles]
+    roles:list[str]
```
When you add remaining tasks to be done, please add one to ensure we write tests for raising  database errors for invalid roles, apps, permissions etc.

sijumoncy replied:
We are testing the role is the global var of roles always and raise error . Will add this test in coming PR

kavitharaju commented on app/schema/schema_auth.py:
```diff
@@ -150,3 +149,76 @@ class UserUpdateResponse(BaseModel):
+    email:str
+    name:str
+    organization:str
+    password:types.SecretStr
```
we are expecting Vachan Admin to store these passwords manually right? And also tokens also, right?

sijumoncy replied:
this is the registration input schema for Apps can be accessed from VA app for registering app , need all these fields to create an app in kratos ?

kavitharaju commented on app/routers/auth_api.py:
```diff
@@ -36,31 +39,32 @@ async def register(register_details:schema_auth.Registration,request: Request,#p
     '''Logout
-    * Loging out will end the expiry of a token even if the time period not expired.
-    * Successful login will return a token for user for a time period'''
+    * Loging out will end the expiry of a token even if the time period not expired.'''
```
If we are to use this API for revoking an APP token, what would be the expected usage?
- user_token of VA or SA in the auth header
- app_key in the query param
Is this right?
What happens then? Isn't the user gonna get logged out instead of the app?
- If we pass the APP token as the auth header, then how can we ensure it is done by authorized user?
 Do we have another API to expire APP token?

sijumoncy replied:
- This end point to revert user token. and if we pass app key in the header as like a user token , I believe will raise an error for unexpected token type .

- App key in the query params is like all other apis to pass the app token from which request is coming.
- The App key revert have another endpoint currently restricted to acces only from VA app and VA, SA can perform the revert operation as per the MAP and Rules

kavitharaju commented on db/csvs/auth_resource_types.csv:
```diff
@@ -0,0 +1,19 @@
+resource_type_name,resource_type_description
```
I think the file could be named `api_resource_types.csv`

kavitharaju commented on docker/Kratos_config/email-password/kratos-app.yml:
```diff
@@ -0,0 +1,88 @@
+
+serve:
+  public:
+    base_url: http://kratos:4443/
```
This and the below one are the container names we discussed about changing.

sijumoncy replied:
yes , I will change those, check and Inform

kavitharaju commented on docker/Kratos_config_App/database.yml:
```diff
@@ -0,0 +1,117 @@
+  kratos-migrate:
+    image: oryd/kratos:v0.7.0-alpha.1
+    environment:
+      - DSN=${VACHAN_APP_AUTH_DATABASE:-postgres://kratos:secret@postgresd:5434/kratos?sslmode=disable&max_conns=20&max_idle_conns=4}
```
I think it would be worth taking a look at the port numbers of the Kratos postgres DBs. Here I see 5434. Some where else I saw 5435. Better stick to one on all places, quickstart, database and test-dependencies to avoid confusion

sijumoncy replied:
35 is in the local quickstart.yml.

kavitharaju commented on docker/Kratos_config_App/email-password/kratos.yml:
```diff
@@ -0,0 +1,88 @@
+version: v0.4.6-alpha.1
```
Is this version correct?

sijumoncy replied:
this need to be changed to latest 10.1 , https://github.com/Bridgeconn/vachan-api/issues/411 mentioned for yml fixes

kavitharaju commented on docker/Kratos_config_App/database.yml:
```diff
@@ -0,0 +1,117 @@
+version: "3.7"
```
Why do we have both folders `docker/Kratos_config` and `docker/Kratos_config_App` with similar files repeated? Can't we just keep one?

sijumoncy replied:
Kratos_config App folder is the one method that i tried earlier for running seperate instance of kratos , that we discussed , kept this for a reference only , will remove once the current setup start working smooth

kavitharaju commented on docker/run-tests.yml:
```diff
@@ -1,37 +1,39 @@
+      - VACHAN_KRATOS_PUBLIC_URL=http://kratos:4433/
+      - VACHAN_KRATOS_ADMIN_URL=http://kratos:4434/
+      - VACHAN_KRATOS_APP_PUBLIC_URL=http://kratos:4443/
+      - VACHAN_KRATOS_APP_ADMIN_URL=http://kratos:4444/
```
Are both set of URLs connecting to same "kratos" container here? Should the app ones be different?

sijumoncy replied:
https://github.com/Bridgeconn/vachan-api/issues/411 yes will change it to kratos-app. right ?

kavitharaju replied:
yes.

sijumoncy commented:
- [x] Removed API user from default apps ( added only for tests )
- [x] GQL test disable using pylint command in run test yml
- [x] role table updated with app and description column
- [x] refactored app , user methods
- [x] yml files updated
- [x] filter apps enum removed
- [x] bugs and issue fixed in PR comments
- [x] all test fixed other than 1 test case failing for SW

kavitharaju reviewed (approved):
Lets proceed with this as we have our tests working now.
Once all APIs are in place we will be able to test them better with automated tests as well as a round of manual testing
